libutils/array_map.c

```diff
@@ -52,8 +52,12 @@ int ArrayMapInsert(ArrayMap *map, void *key, void *value)
     {
         if (map->equal_fn(map->values[i].key, key))
         {
-            map->destroy_key_fn(key);
+            /* Replace the key with the new one despite those two being the
+             * same, since the new key might be referenced somewhere inside
+             * the new value. */
+            map->destroy_key_fn(map->values[i].key);
             map->destroy_value_fn(map->values[i].value);
+            map->values[i].key   = key;
```

Comment:
Well there was a very mysterious crash that was resolved by accident after changing this. :-p